### ISSUE TITLE
Mediated Holder Exchanges

### ISSUE THREAD

**Issue by troyronda** [w3c-ccg/universal-wallet-interop-spec#84]
Cross-posted from https://github.com/w3c-ccg/vc-http-api/issues/176 as I mentioned universal wallet.

**Comment by OR13** [comment]
I am excited to see this added to the universal wallet interop spec, we've been working on VP Request Spec JSON based APIs for exchanging content.

I see capabilities to a mediator a natural next step... summarizing:

1. Notify a VP (receive a domain and challenge)
2. Submit a VP (sign over the domain and challenge)

Consider the case where the VP contains an EDV authorization capability:

Receiving party can now invoke the capability and access all data associated with the requestors flow.

You can see a similar proposal we made here: https://github.com/decentralized-identity/confidential-storage/issues/157

**Comment by OR13** [comment]
Here is a snippet from our universal wallet plugin for vp exchange:

```
it("alice notifies bob of intent to present", async () => {
  // request
  m0 = aliceWallet.createNotificationQueryRequest(
    "IntentToSellProductCategory"
  );
  // alice.send(m0, bob)

  // response
  m1 = bobWallet.createNotificationQueryResponse("bob.example.com", m0);
  // bob.send(m1, alice)
});

it("alice responds to bobs challenge and query with a vp", async () => {
  // normally alice might have this vc, in this case
  // for testing here, we create it on the fly
  m2 = await getVp(m1);
  // alice.send(m2, bob)
});

it("bob verifies and stores alice's vp", async () => {
  await bobWallet.verifyAndAddPresentation(m2, {
    suite: new Ed25519Signature2018(),
    documentLoader: documentLoader,
  });
  expect(bobWallet.contents[2].type).toBe("FlaggedForReview");
});
```

**Comment by OR13** [comment]
we're still experimenting with the interfaces, but we'll be making sure to align them in the spec here, based on feedback we receive.

**Comment by OR13** [comment]
ping @sudeshrshetty @dmitrizagidulin

**Comment by sudeshrshetty** [comment]
@OR13 I guess you are were designing interfaces for [credential exchange APIs](https://github.com/w3c-ccg/vc-http-api/pull/168) for holders as part of #75 & #76 .

In this mediator holder exchange [flow](https://user-images.githubusercontent.com/1906596/113071546-f6650380-9192-11eb-89b6-e696d0c9a4db.png), I believe only wallet holder comes under universal wallet's scope. 
Universal wallet interfaces might need to cover these steps:

- pushing encrypted presentation to mediator holder.
- returning presentation to verifier containing authorization code & mediator URL .
- registering mediator holders.
- registering service holders.
- for encrypting presentation we can make use of [`query.recipients`](https://w3c-ccg.github.io/vp-request-spec/#format) from credential request.

also we need to consider the fact that a verifier may or may not have infrastructure to pull presentation at later stage from mediator.

**Comment by OR13** [comment]
I'm struggling to grok this exactly.

It feels like there are a couple things, which each might be connected to wallet apis:

## Simple Presentations initiated by the sender pushing

1. wallets can present to each other over http. ( VC-HTTP-API / Traceability API / VP Request Spec )
2. wallets can present to websites ( CHAPI / VP Request Spec ) 

## Mediated Presentations initiated by the receiver pulling

<img width="735" alt="Screen Shot 2021-04-15 at 5 03 28 PM" src="https://user-images.githubusercontent.com/8295856/114943999-826e6080-9e0c-11eb-840d-fabad65476f9.png">

In both cases, there is some setup / registration that the wallet likely needs to have done.

Here you can see the setup we consider for the simple push oriented presentations:

- https://github.com/transmute-industries/verifiable-data/blob/main/packages/universal-wallet-vp-exchange-plugin/src/__tests__/vp-exchange.spec.ts
- https://github.com/OR13/simple-presentations

Can we perhaps label the steps in mediated presentations in terms of vp request spec and edv's?

**Comment by OR13** [comment]
@troyronda @sudeshrshetty 

There are arrows with no labels, but I will try to add labels to the stuff that is numbered.

1. VP-Request Spec, Alice Requests Credentials from Bob.
2. Bob's front end wallet, tells his back end wallet (Service Holder?)... to construct and store a VP 
3. Bob's Backend wallet (Service Holder?) pushes the VP to Mediator EDV.

...

I am stuck, I don't see how step 4 is triggered by the verifier (there is no response to them to step 1)

and EDVs don't push data, so the verifier is left in a state of waiting...

**Comment by OR13** [comment]
### Case 0

<img width="978" alt="Screen Shot 2021-04-16 at 10 44 02 AM" src="https://user-images.githubusercontent.com/8295856/115049746-ab403580-9ea0-11eb-96f3-52f1cd8f227e.png">

In this flow, the holder user agent wallet has everything, and there is not "backend holder wallet".

**Comment by OR13** [comment]
### Case 1
<img width="942" alt="Screen Shot 2021-04-16 at 10 54 46 AM" src="https://user-images.githubusercontent.com/8295856/115051198-2bb36600-9ea2-11eb-8246-d01540010c9d.png">

Is like case 0, but instead of directly submitting the encrypted VP to the mediator, the holder user agent requests a backend service do so.

**Comment by OR13** [comment]
Both case 0 and case 1, assume a verifier can decrypt.... this requires the verifier to do key management.

If we remove this requirement, the mediator can observe the presentation content.

For the sake of completeness, here is the original exchange API proposal (the one the traceability api supports today).

<img width="947" alt="Screen Shot 2021-04-16 at 11 07 01 AM" src="https://user-images.githubusercontent.com/8295856/115052831-058ec580-9ea4-11eb-97e2-bbef5a8f625c.png">

**Comment by OR13** [comment]
Personally,  I prefer case 2 as a starting point, not just because its what we support today, but because we have no real need for yet another word for holder... a mediator is a holder who aggregates presentations they can't even read, and then submits them as an unsigned presentation.... mediators don't do encryption or signing and they don't need to since they can't cheat if the verifier supports encryption.

For folks wondering how CHAPI fits into all this, CHAPI uses an iframe as a mediator... 

https://github.com/digitalbazaar/credential-handler-polyfill/blob/0395b0c51f25bab2c0b2e727a44df0932dde5142/index.js#L18

Without encryption, which is safe as long as the polyfill isn't compromised.

<img width="938" alt="Screen Shot 2021-04-16 at 11 19 56 AM" src="https://user-images.githubusercontent.com/8295856/115054286-b053b380-9ea5-11eb-9c85-3960390bea7c.png">

There are 2 show stoppers in this flow.

1. Verifier Front End
2. Mediator iFrame

Thats pretty much game over for mobile application and machine to machine use cases for presentation exchange.

**Comment by troyronda** [comment]
> Personally, I prefer case 2 as a starting point.

For us, the case where we encrypt the VP through the back-end is quite important. We also have cases where we want to aggregate multiple VPs prior to the Verifier pulling. Agree that this has the complexity where the Verifier needs to decrypt. Of course, they already need to verify signatures so they have crypto work to do regardless.

> a mediator is a holder

Yes - I think we should view this component as a Holder.

**Comment by OR13** [comment]
If case 2, step 6 used an encrypted VP, its essentially the same exact flow as case 0.

**Comment by OR13** [comment]
It seems like the man things missing from the universal wallet spec to support these flows are:

1. JSON models for queries, with mediator addresses, and encryption keys (partially described in VP request spec)
1. An encrypted presentation object format, compatible with JWE.

**Comment by troyronda** [comment]
Another one:

In these "non-interactive" cases — like user agent creating the VP and/or asking services to create the VP, I assumed domain and challenge goes with Step 1. (and of course, in the final verifier pulls step, the back-end holder can have additional controls regarding domain before supplying the result).

**Comment by OR13** [comment]
yes, if the domain and challenge are part of the initial query, we can skip case 2 step 4, and got right to step 5.

**Comment by OR13** [comment]
<img width="1127" alt="Screen Shot 2021-04-16 at 2 36 26 PM" src="https://user-images.githubusercontent.com/8295856/115075720-23b6ee80-9ec1-11eb-8245-82e1295ee703.png">

^ If we assume verifiers can decrypt.

**Comment by OR13** [comment]
in case 2.1, the verifier an receive an unsigned VP directly from the backend holder wallet... which in turn may hav collected VCs and VPs from an unknown number of intermediaries, before encrypting the aggregated VPs.... while its possible for 5 to use a signed VP, its not necessary, since each encrypted VP inside may be signed.

And the verifier will need to decrypt and verify each VP anyway.

**Comment by sudeshrshetty** [comment]
@OR13 about cases with mediator involved.

in some scenarios where a query can produce multiple credentials and we may have case 0 & case 1 combined. In that case, mediator has to wait for service holder to finish generating encrypted presentation. How mediator will know to wait for service holder before pushing notification to verifier?

**Comment by OR13** [comment]
@sudeshrshetty lets look at case 2.1 for the answer to that,

in step 4, the service agent gathers all VPs necessary to reply to the initial query... those VPs might come from the User Agent, or the might come from other backend services that the Service Agent as access too.


In either case, the service agent has the query, and knows then they have enough to proceed to step 5.

**Comment by troyronda** [comment]
> those VPs might come from the User Agent

Yup - Part of "Collect Encrypted VPs" needs to allow a User Agent to push encrypted VPs to the Service Agent.

**Comment by sudeshrshetty** [comment]
Consolidating all the points from case 2.1, 

- **1 verifier sending query to user agent holder**:  Any query from existing vp-request-spec can be supported. [`query.recipient`](https://w3c-ccg.github.io/vp-request-spec/#format) can be used to supply encryption keys. In case of CHAPI, a response can be returned to notify verifier that request is successfully accepted.
- **2 notify await mediator**
- **3 push query**: user agent holder can push query or presentation (from locally found credentials) or both(for partially found results) to service agent holder. user agent may also need to send authorization (ex: zcap headers) to authorize service agent for pulling presentations from backend services (or agents).  User agent may also need to forward encryption keys it received from verifier to be used by backend holders. A service agent can also be customized to act as mediator holder.
- **4 backend holder wallet gathers VCs and VPs**: Service holder can call `n` number of backed services(or holder agents) to collect the results.
- **5 push unsigned presentations of encrypted signed presentations over domain and challenge**:  Can also have an unbounded number of untrusted mediators here.
-  **6/7 presentation exchange complete**: 

@OR13 @troyronda let me know if I have misinterpreted or missed something.

Here are some assumptions:

- All backend holders including service holder are already registered to user agent holder. To enable user agent holder to choose right backend holder for incoming query, some kind of metadata exchange should happen during registration between user agent holder and backend holder. OR all backend holders can get registered to service holder instead to hide complexity from user agent holder.
- In case of multiple simultaneous queries sent from verifier front end to user agent holder: domain and challenge are enough to correlate response from verifier backend OR user agent holder should return some kind of token as a response to **step 1** for correlation.

**Comment by OR13** [comment]
I'm at a session in IIW reviewing https://specs.bloom.co/wallet-and-credential-interactions/

**Comment by OR13** [comment]
@sudeshrshetty I think you got all the points... the main one we've been struggling with is how to support the registration phase, and how it relates to discovery...

I think naively, we assume from a public organization registry, like BCGov org book, a holder discovers credentials they may receive or present, does some setup phase, maybe related to "connections"... and then the protocols we are talking about here kickoff.

**Comment by sudeshrshetty** [comment]
@OR13 how about using [credential manifest](https://identity.foundation/credential-manifest/) during registration of backend holder with user agent holder? Backend holders can send credential information in similar data format during registration.

**Comment by OR13** [comment]
@sudeshrshetty yeah, I think its possible that we might move faster building on top of the DIF / Bloom specs in this regard.
